### PR TITLE
Stabilise code, add API docs, and prepare for 1.0 release

### LINKED ISSUE #449: Remove all automatic conversion of names of key/value of case classes, case objects and sealed traits
zio-config inherited the same mistake as most of the other libraries in Scala in assuming the name of the caseClass or sealed trait names as snakeCase/kebab_case. These assumptions are removed.

For any custom names, there are various of doing it in zio-config
Example:
```
sealed trait Region

@name("ap-southeast-2")
case object ApSouthEast2 extends Region

```

### PR DESCRIPTION
**Pull request by afsalthaj**
There will be more commits on this PR. Early feedbacks will be important for this PR.
Note: Load large difference explicitly in PR

Also fix  #449

### CODE REVIEW

**Comment by afsalthaj** [comment]
@jdegoes  Could you take a look at it?

File: core/src/main/scala/zio/config/ConfigDescriptorModule.scala [hunk 1]
```diff
@@ -6,230 +6,1311 @@ trait ConfigDescriptorModule extends ConfigSourceModule { module =>
...
+     *
+     * However, the reverse relationship (which is used to retrieve
+     * the original type from its transformed representation) is `String => Int` is
+     * not a total function, unless it was `String => Either[error, Int]`.
+     *
+     * That is, Not all elements of set `String` can be converted to `Int`.
+     * Example: A string "abc" cannot be converted to `Int`.
+     *
+     * Hence we can do `s => Try(s.toInt).toOption` to mark the possibility of errors.
+     * This is a function of the type: `B => Option[A]`.
```
Comment (afsalthaj) [inline core/src/main/scala/zio/config/ConfigDescriptorModule.scala:42]: Remove lines 34 to 42

File: core/src/main/scala/zio/config/ConfigDescriptorModule.scala [hunk 2]
```diff
@@ -6,230 +6,1311 @@ trait ConfigDescriptorModule extends ConfigSourceModule { module =>
...
+          case OrElse(value1, value2) =>
+            OrElse(value1.map(loop), value2.map(loop))
+        }
 
       loop(self)
     }
 
+    /**
+     * `optional` function allows us to tag a configuration parameter as optional.
+     * It implies, even if it's missing configuration will be a success.
```
Comment (afsalthaj) [inline core/src/main/scala/zio/config/ConfigDescriptorModule.scala:495]: `configuration` to `retrieving configuration`

File: core/src/main/scala/zio/config/ConfigDescriptorModule.scala [hunk 3]
```diff
@@ -6,230 +6,1311 @@ trait ConfigDescriptorModule extends ConfigSourceModule { module =>
...
+     *     final case class Config(userName: String, port: Option[Int])
+     *
+     *     object Config {
+     *        val dbConfig: ConfigDescriptor[Config] =
+     *           (string("USERNAME") |@| int("PORT").optional)(Config.apply, Config.unapply)
+     *     }
+     *  }}}
+     *
+     *
+     *   The fact that it is an optional in error messages if config retrieval
```
Comment (afsalthaj) [inline core/src/main/scala/zio/config/ConfigDescriptorModule.scala:517]: Reword.

Comment (afsalthaj) [reply core/src/main/scala/zio/config/ConfigDescriptorModule.scala:517]: Or probably remove it

**Review by jdegoes — DISMISSED** [review]
Looks good, let's get it in!

**Review by jdegoes — APPROVED** [review]
Looks good although `lint` still fails. Good to merge when it passes!

**Comment by afsalthaj** [comment]
@jdegoes  Sorry that I am updating more changes after our last discussion, and its almost done.
I will need your review after more api docs are updated.

**Review by jdegoes — APPROVED** [review]

**Pull request merged**